source/util/FeatureRegistry.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "FeatureRegistry.h"
+#include "../logging/LoggerFactory.h"
+
+using namespace Aws::Iot::DeviceClient;
+using namespace Aws::Iot::DeviceClient::Util;
+using namespace Aws::Iot::DeviceClient::Logging;
+
+constexpr char FeatureRegistry::TAG[];
+
+std::shared_ptr<Feature> FeatureRegistry::get(const std::string &name) const
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    for (auto &feature : features)
+    {
+        if (feature->getName() == name)
+        {
+            return feature;
+        }
+    }
+
+    LOGM_WARN(TAG, "Feature, %s, not found in current registry", name.c_str());
+
+    return nullptr;
+}
+
+void FeatureRegistry::add(std::shared_ptr<Feature> feature)
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    if (std::find(features.begin(), features.end(), feature) != features.end())
+    {
+        LOGM_WARN(TAG, "Attemped to add %s to Feature Registry despite already existing", feature->getName().c_str());
+    }
+    else
+    {
+        features.push_back(feature);
+    }
+}
+
+void FeatureRegistry::remove(const std::string &name)
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    for (int i = 0; (unsigned)i < features.size(); i++)
+    {
+        if (features.at(i)->getName() == name)
+        {
+            features.erase(features.begin() + i);
+            break;
+        }
+    }
+}
+
+int FeatureRegistry::getSize()
```

Comment:
1. Return type of `std::vector::size` is `std::size_t` (or `std::vector::size_type` which is same thing). Similar to the previous comment, use `std::size_t` instead of `int`.
2. Since this class maintains a lock to protect concurrent modification of `features`, then this function also needs to acquire a lock before invoking member function on `features`.